tests/unit/test_charm.py

```diff
@@ -533,3 +538,81 @@ def test_publish_snapshot_action_fail(
         self.harness.charm._get_snapshot_name = Mock()
         self.harness.charm._on_publish_snapshot_action(action_event)
         self.assertEqual(action_event.fail.call_args, call("Snapshot does not exist"))
+
+    def test_list_snapshot(self):
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            base_path = pathlib.Path(tmpdirname)
+            with patch("builtins.open", new_callable=mock_open):
+                self.harness.update_config(
+                    {
+                        "base-path": str(base_path),
+                    }
+                )
+            expected_snapshots = [
+                base_path / "snapshot-1970010{}".format(i) for i in range(3)
+            ]
+            for snapshot in expected_snapshots:
+                snapshot.mkdir(parents=True)
+            returned_snapshots = self.harness.charm._list_snapshots()
+            self.assertEqual(sorted(expected_snapshots), sorted(returned_snapshots))
+
+    def test_get_snapshot_name(self):
+        snapshot_name = self.harness.charm._get_snapshot_name()
+        part_0 = snapshot_name.split("-")[0]
+        part_1 = snapshot_name.split("-")[1]
+        self.assertEqual(part_0, "snapshot")
+        self.assertTrue(datetime.datetime.strptime(part_1, "%Y%m%d%H%M%S"))
+
+    def test_check_packages_action(self):
+        freed_up_space = "0.0 bytes"
+        packages_to_be_removed = [Mock() for i in range(3)]
+
+        self.harness.charm._check_packages = Mock()
+        self.harness.charm._check_packages.return_value = [
+            packages_to_be_removed,
+            freed_up_space,
+        ]
+        action_event = Mock()
+        self.harness.charm._on_check_packages_action(action_event)
+        self.assertEqual(
+            action_event.set_results.call_args,
+            call(
+                {
+                    "message": (
```

Comment:
It doesn't make sense to limit the message in unit test to me.
Unit test is used to verify the specification of the function.
I would like to leave some flexibility for the implementation unless the format of message is part of specification and it will impact the caller.

assert_called_once may be good enough.

Reply:
Thanks, that's indeed true. I will update that.